code/datums/progressbar.dm

```diff
@@ -2,119 +2,228 @@
 #define PROGRESSBAR_ANIMATION_TIME 5
 
 /datum/progressbar
-	var/max_progress
-	var/listindex
+	///The progress bar visual element.
+	var/image/bar
+	///The progress bar visual element - but for the target.
+	var/image/targetbar
+	///The target where this progress bar is applied and where it is shown.
+	var/atom/target
+	///Whether the target should be shown a bar as well
+	var/show_target
+	///The target's client, if applicable. Only set if the target is intended to see the progress bar as well.
+	var/client/target_client
+	///The mob whose client sees the progress bar.
+	var/mob/user
+	///The client seeing the progress bar.
+	var/client/user_client
+	///Effectively the number of steps the progress bar will need to do before reaching completion.
+	var/goal = 1
+	///Control check to see if the progress was interrupted before reaching its goal.
+	var/last_progress = 0
+	///Variable to ensure smooth visual stacking on multiple progress bars.
+	var/client_listindex = 0
+	///Variable to ensure smooth visual stacking on multiple progress bars - but for the target.
+	var/target_listindex = 0
+	///An optional, clickable object that can be used to speed up progress bars
+	var/obj/screen/progressbar_booster/booster
+	///How much bonus progress we've accured from a linked progress booster
+	var/bonus_progress = 0
+
+/datum/progressbar/New(mob/_user, goal_number, atom/_target, _show_target = FALSE, bonus_percentage, focus_sound, focus_frequency)
+	. = ..()
+	if (!istype(_target))
+		CRASH("Invalid target [_target] passed in")
+	if(QDELETED(_user) || !istype(_user))
+		CRASH("a /datum/progressbar was created with [isnull(_user) ? "null" : "invalid"] user")
+	if(!isnum(goal_number))
+		CRASH("a /datum/progressbar was created with [isnull(goal_number) ? "null" : "invalid"] goal_number")
```

Comment:
replace crash() with stack_trace, we have something like auxtools_stack_trace or crash_with (literally same thing) and add back qdel(src)

Reply:
done